### PR TITLE
Bugfix/all gather

### LINKED ISSUE #5181: all_gather raises NotImplementedError when no Accelerator defined in Trainer
## 🐛 Bug

When no Accelerator is defined in Trainer, all_gather function in LightningModule raises NotImplementedError.

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1VPEIaQ-aN5KVA70VtvGk24AVkTYPvMoY?usp=sharing

### Expected behavior

* All gather should return a value instead.

### Environment

 * CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.4
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.2rc1
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #5181

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -203,10 +204,13 @@ def all_gather_ddp_if_available(
     Return:
         A tensor of shape (world_size, batch, ...)
     """
+    if group is None:
+        group = torch.distributed.group.WORLD
```
Comment (SkafteNicki) [inline pytorch_lightning/utilities/distributed.py:208]: same change as this PR https://github.com/PyTorchLightning/pytorch-lightning/pull/5125

**Comment by SkafteNicki** [comment]
@tchaton should this PR target master branch?

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -203,10 +204,13 @@ def all_gather_ddp_if_available(
...
+
     if torch.distributed.is_available() and torch.distributed.is_initialized():
         if sync_grads:
             return AllGatherGrad.apply(tensor, group)
         else:
-            with torch.no_grad:
+            with torch.no_grad():
```
Comment (awaelchli) [inline pytorch_lightning/utilities/distributed.py:214]: wasn't this already merged?

**Comment by tchaton** [comment]
Release/1.2-dev is not up-to-date with master.  And we decided we won't merge any new features to master during Christmas holiday as it creates a potential risk.

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py:219]: Yes, it was but on master. This new release system might be problematic in the future. Let's see how it goes.

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -203,10 +204,18 @@ def all_gather_ddp_if_available(
     Return:
         A tensor of shape (world_size, batch, ...)
     """
+    if group is None:
+        group = globals()["group"].WORLD
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/distributed.py:208]: Why do we do this over just `group.WORLD`?

Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py]: removed

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -363,7 +366,10 @@ def __auto_choose_log_on_epoch(self, on_epoch):
 
         return on_epoch
 
-    def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False):
+    def all_gather(self,
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:369]: let's use the standard formating...

File: pytorch_lightning/utilities/distributed.py [hunk 5]
```diff
@@ -203,10 +204,18 @@ def all_gather_ddp_if_available(
     Return:
         A tensor of shape (world_size, batch, ...)
     """
+    if group is None:
+        group = torch.distributed.group.WORLD
+        if group is None:
+            raise MisconfigurationException(
+                "The provided group was None and `torch.distributed.group` isn't available. "
+                "Gathering tensor across processes won't be possible. "
```
Comment (Borda) [inline pytorch_lightning/utilities/distributed.py:212]:
```suggestion
                "The provided group was None and `torch.distributed.group` isn't available."
                " Gathering tensor across processes won't be possible."
```

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -379,7 +387,18 @@ def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, s
         Return:
             A tensor of shape (world_size, batch, ...)
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:388]:
```suggestion
            A tensor of shape (world_size, batch, ...), or if the input was a collection the output will also be a collection with tensors of this shape.
```

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:388]: the rest of the docs above need to be updated with the renaming you did. 
in fact, sphinx should warn about this and I'm not sure why docs build is passing on this.

File: tests/utilities/test_all_gather_grad.py [hunk 7]
```diff
@@ -41,3 +44,48 @@ def _test_all_gather_ddp(rank, world_size):
 def test_all_gather_ddp():
     world_size = 3
     torch.multiprocessing.spawn(_test_all_gather_ddp, args=(world_size,), nprocs=world_size)
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="DDP not available on windows")
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.skipif(not os.getenv("PL_RUNNING_SPECIAL_TESTS", '0') == '1',
+                    reason="test should be run outside of pytest")
+def test_all_gather_properly_works(tmpdir):
```
Comment (awaelchli) [inline tests/utilities/test_all_gather_grad.py:53]:
```suggestion
def test_all_gather_collection(tmpdir):
```
test could have a better name as it is obvious that things should work properly x))

**Review by awaelchli — COMMENTED** [review]
looks ok but need to investigate gpu failure (drone)

File: pytorch_lightning/core/lightning.py [hunk 8]
```diff
@@ -373,14 +381,27 @@ def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, s
...
+        if self.trainer.accelerator_backend is not None:
+            all_gather = self.trainer.accelerator_backend.all_gather
+        else:
+            all_gather = all_gather_ddp_if_available
+
+        def to_dtype_tensor(value, dtype=None):
+            return torch.tensor(value, dtype=dtype, device=self.device)
+
+        data = apply_to_collection(data, float, partial(to_dtype_tensor, dtype=torch.float))
+        data = apply_to_collection(data, int, partial(to_dtype_tensor, dtype=torch.int))
```
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:402]: Do you know if torch.long/half are supported?

File: pytorch_lightning/utilities/distributed.py [hunk 9]
```diff
@@ -202,6 +204,14 @@ def all_gather_ddp_if_available(
     Return:
         A tensor of shape (world_size, batch, ...)
     """
+    if group is None:
+        group = torch.distributed.group.WORLD
+        if group is None:
+            raise MisconfigurationException(
+                "The provided group was None and `torch.distributed.group` isn't available."
+                " Gathering tensor across processes won't be possible."
+            )
```
Comment (justusschock) [inline pytorch_lightning/utilities/distributed.py:213]: Is this really possible/necessary? IMO this would be catched by the below statement (`if torch.distributed.is_available() and torch.distributed.is_initialized()`) if you move the above statement (`if group is None: group = torch.distributed.group.WORLD` to the if clauses scope.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py]: removed !

File: pytorch_lightning/core/lightning.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py]: add support for bool and np.ndarray.

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 10]
```diff
@@ -373,14 +381,23 @@ def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, s
         distributed processes
 
         Args:
-            tensor: tensor of shape (batch, ...)
+            tensor: int, float, tensor of shape (batch, ...),  or a collection of
+                int, float, tensor of shape (batch, ...)
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:385]:
```suggestion
            tensor: int, float, tensor of shape (batch, ...), or a (possibly nested) collection thereof.
```
it can be nested arbitrarily, right?

File: pytorch_lightning/utilities/apply_func.py [hunk 11]
```diff
@@ -27,11 +30,18 @@
     Batch = type(None)
 
 
+CONVERSION_DTYPES = [
+    (bool, torch.bool),
+    (int, torch.int),
+    (float, torch.float),
+    (np.ndarray, None),
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:37]: alternatively you could define a callable (torch.from_numpy) here instead of specifically checking for ndarray below

**Review by awaelchli — APPROVED** [review]
useful!

**Review by williamFalcon — APPROVED** [review]
